client/storage/index-new.js

```diff
@@ -0,0 +1,436 @@
+        if (!chunkInfoString.startsWith(CHUNKINFO_PROLOGUE)) {
+            logger.debug(`File ${file.id} consists of a single chunk, returning it`);
+            await fs.writeFile(filePath, chunkInfo);
+
+            file.size = chunkInfo.length;
+            file.dl_status = DOWNLOAD_STATUS.COMPLETED;
+            await file.save();
+
+            return encoding === null ? chunkInfo : chunkInfo.toString(encoding);
+        }
+
+        logger.debug(`Processing chunk info, file id: ${file.id}`);
+
+        const {
+            type,
+            hash,
+            chunks,
+            filesize,
+            merkle: merkleHash
+        } = JSON.parse(chunkInfo.slice(CHUNKINFO_PROLOGUE.length + 1));
+
+        if (type !== 'file') {
+            throw new Error('Bad file type');
+        }
+
+        if (hash !== 'keccak256') {
+            throw new Error("Bad hash type");
+        }
+        const merkleReassembled = merkle
+            .merkle(chunks.map(x => Buffer.from(x, 'hex')), hashFn)
+            .map(x => x.toString('hex'));
+        if (!areScalarArraysEqual(merkleReassembled, merkleHash)) {
+            throw new Error("Incorrect Merkle hash");
+        }
+
+        logger.debug(`Chunk info for file id ${file.id} processed, getting chunks`);
+
+        // TODO: retry logic
+        const chunkBuffers = (await Promise.all(chunks.map(chunkId => getChunk(chunkId, encoding))));
+        const fileBuffer = Buffer.concat([
+            ...chunkBuffers.slice(0, -1),
+            // We should trim the trailing zeros from the last chunk
+            chunkBuffers[chunkBuffers.length - 1]
+                .slice(0, filesize - ((chunkBuffers.length - 1) * config.chunk_size_bytes))
+        ]);
+        
+        logger.debug(`Successfully proceeded file chunks, file id: ${file.id}`);
+
+        await fs.writeFile(filePath, fileBuffer);
+
+        file.size = filesize;
+        file.dl_status = DOWNLOAD_STATUS.COMPLETED;
+        await file.save();
+
+        return encoding === null ? fileBuffer : fileBuffer.toString(encoding);
+    } catch (e) {
+        logger.error(`File download failed, file id: ${file.id}`);
+        logger.error(e.stack);
+        file.dl_status = DOWNLOAD_STATUS.FAILED;
+        await file.save();
+        throw e;
+    }
+};
+
+const getJSON = async (id, useCache = true) => {
+    logger.debug(`Getting JSON ${id}`);
+    const file = await getFile(id, "utf8", useCache);
+    return JSON.parse(file.toString("utf-8"));
+};
+
+const getFileIdByPath = async (dirId, filePath) => {
+    const directory = await getJSON(dirId);
+    const segments = filePath.split(/[/\\]/).filter(s => s !== "");
+    const nextFileOrDir = directory.files.find(f => f.name === segments[0]);
+    if (!nextFileOrDir) {
+        throw new Error(`Failed to find file ${filePath} in directory ${dirId}: not found`);
+    }
+    if (segments.length === 1) {
+        return nextFileOrDir.id;
+    }
+    if (nextFileOrDir.type === "fileptr") {
```

Comment:
Let's add type enum in order to avoid typos.

Reply:
Done